Introduce support for Tablets

Lorak-mmk opened the pull request:
This PR introduces support for tablets. The design as a bit different than in the other drivers.

### For information about what Tablets are see:
https://www.scylladb.com/2023/07/10/why-scylladb-is-moving-to-a-new-replication-algorithm-tablets/

### Driver context
To give a bit of context of how things are working in Rust Driver now:
`Session` has a `cluster: Cluster` field which in turn has `data: Arc<ArcSwap<ClusterData>>`.
`ClusterData` contains all the schema and topology information. It also contains `ReplicaLocator`, which uses token ring to return a list of replicas for a given token and replication strategy.
Driver can access `ClusterData` cheaply, without the need for any locks. It is used for performing queries and is also exposed to the user: https://docs.rs/scylla/latest/scylla/transport/session/struct.Session.html#method.get_cluster_data
This data is updated by `ClusterWorker`, running in a separate Tokio task. This worker handles server events, `use_keyspace` requests, manages control connection. When necessary (and also periodically) it re-fetches topology and schema
information, creates new `ClusterData` object and then swaps it atomically (for more information see https://docs.rs/arc-swap/latest/arc_swap/ ) with the old one - this way new queries will use new `ClusterData` without disrupting old ones.

### How does the driver interact with Tablets?
Information about Tablets are stored in `system.tablets`. Driver could fetch this information to have it always complete and up to date, but that could be overwhelming for the cluster.
This is because there is just one token ring - token ownership is tied to a node, but in tablets this information is per-table (my intuition about tablets is to see them as per-table token ring that can by dynamically changed). Also there may be more tablets than token ring elements.
Now imagine a scenario where a lot of clients try to reconnect to the cluster, which has a lot of tables. The amount of transferred data would be too big.
For this reason driver fetches tablets information lazily. When the driver send a statement to a non-replica node, it will receive tablet information in the custom payload of the response.
This information contains data about the tablet that the statement belongs to: first and last tokens of the tablet + list of tablet replicas (in the form `list<(uuid, shard)>` ).
Driver can then update it's internal tablet information using this feedback, so that the following statements will be sent to a correct replica.

### Implementation

Information about tablets is stored in `ReplicaLocator` (which is inside `ClusterData`), similarly to token ring data.
When a connection receives tablet feedback it sends this feedback trough a channel. Receiving end of this channel
is owned by `ClusterWorker`. When it receives something on this channel it clones `ClusterData` (because there is no other way to update it without introducing locks),
updates tablet info inside and stores it in `ArcSwap` - just like when topology is updated. As an optimisation it tries to retrieve several tablets from the channel at once,
so that when there is a lot of updates we don't perform clone + swap for each of them.

This way queries can be executed as before, without any locks and other costly operations, just accessing `ClusterData` which is mostly a simple data structure.

### Maintanance

Implementation chosen by Scylla makes things a bit more complicated for the driver.  We only update tablet info when we send a query to a wrong node.
This means that the information can sometimes become stale
- When a node is removed we would just keep sending queries to other replicas, never receiving new tablet.
- When a keyspace / table is dropped we would unnecessarily keep info about it.
- (may be Rust Driver only) When driver is not aware of one of the replicas yet (e.g. because the node is new and driver didn't yet fetch topology) and we skip this replica we'll increase load on other replicas indefinitely.
- (may be Rust Driver only) When a `Node` object is recreated (this happens when node changes Ip / DC / Rack) tablet info would still contain old objects.

To handle those scenarios driver fixes tablet information after fetching new topology. As far as I'm aware other driver don't handle those scenarios, we may need to fix it.
One scenario which I'm not sure how to handle (and if it's even problematic): what if user changes RF of a keyspace? How does it work with tablets?
If it just adds a replica to replica list, then we will not use this replica - we'll keep sending requests to other replicas and never receive new tablet info.
I'm not sure how we could detect it and if it's even a problem.

### Alternative solution

One alternative solution would be to use `RwLock` - take a read lock in `ReplicaLocator` to get list of replicas, take a write lock when handling custom payload tablet feedback to update info.
I decided against it because:
- It's possible we would hold a write lock often, slowing down queries. I also don't like the idea of taking a lock in a hot path (executing a query) in general.
- Most of `ClusterData` (pretty much all of it except `Node` objects) is a simple data, without any locks, which makes it simpler to reason about it. I didn't want to change this.
- It makes implementing tablet info maintenance much harder. Doing maintenance would require taking a write lock for a lot of time, hindering performance.

### Downsides of implemented solution

- We need to copy `ClusterData` to update tablet info. This won't increase max memory usage - because we already copy it when updating topology - but now we'll copy it more often. Implementing https://github.com/scylladb/scylla-rust-driver/issues/595 will largely mitigate this issue because we won't need to copy schema information, just topology.
- Received tablet info is not immediately used. There may be a brief period after we receive tablet information, but before it is present in `ClusterData` during which we'll keep sending data to wrong nodes. I don't think it's much of a problem - the same thing would happen with any implementation  if you send a few queries to the same parttition at once. The worst thing that happens is that some more queries will be sent to wrong nodes and we'll receive few more feedbacks.

### LWT
One note about LWT: with token ring the LWT optimisation (sending LWT prepared statements to the same node from any clients, to avoid conflicts) was working always. Now it requires having tablet info for a given partition - so a first LWT to a given partition will not use the optimisation.


TODO:
- [x] Proper commit messages
- [x] Verify content of each commit, change when necessary to ensure each passes CI - There are a few unused function warnings.
- [ ] Run SCT tests with this branch
- [ ] Is there any documentation / comments that need updating?
- [ ] ~~Benchmark possible tablets implementations: vector vs map~~
- [ ] Unit tests for maintenance code 
- [x] Don't unwrap when converting number type in Tablet info parsing
- [x] I recently learned that i64::MIN is not a valid token. Need to look into this to check if we properly change it to i64::MAX when hashing. I wonder if such check should instead be moved to `Token` struct.
- [x] CI for tablets, similar to serverless - because it needs unstable scylla version
- [x] Put tablets behind unstable feature flag until they are released in Scylla

## Past questions / considerations

During implementation I had some dillemas questions, mostly solved / answered now, which are outlined below.

### Tests with removed / decommisioned nodes

I'd like to test such scenarios, as suggested by @avelanarius , but I think we don't have a good way to do this now.
There is proxy, but afaik it can only simulate network errors.
It would be nice to use ccm (or something like this) in tests and have Rust API for it to manipulate cluster more easily.
Need to figure out a way to test it now.

Status: I'll run some SCT tests using cql-stress built using this branch of driver.

### Unknown replica

Is it possible to get a replica uuid in custom payload that we don't know about? What should we do in this case?
Right now I'm just ignoring this replica, but I don't like this. This will cause load to be distributed less evenly
because we skip one (or more) of replicas.

Status: Implemented tablet maintanance to guard against this and some other problems.

### What about `ReplicasOrdered`?

It returns replicas in ring order, which makes little sense in Tablets tables.
I'm thinking of just editing a comment to say that the order is just consistent, but doesn't have to be ring order.
`ReplicasOrdered` seems to only be used for LWT, so this should be fine.
To do this, I need to make sure that the order is actually consistent - because it's not obvious.

Also, I'd like to have some test that uses the path where it's used - probably neeed to use LWT with Tablet-enabled table?

Status: Assumed that replicas are returned in consistent order. Added integration tests performing LWT queries.

### Integration tests

I wrote an integration test for tablets (last commit) but it doesn't always work, and I'm not sure how to write a proper test.
The test first inserts some row. Each query is executed multiple times and the test checks that at least one execution
was sent to wrong node and recevied tablet feedback. This is not always the case and I'm not sure how to force it.

@avelanarius suggested that there is `"tablet_allocator_shuffle"` HTTP api, but the comment in Scylla tests says
` # Increases the chance of tablet migration concurrent with schema change` which doesn't sound helpful here.

In Python driver implementation the test looks into tracing to verify that the query was sent to a correct node.
I'm not a big fan of this approach, as it doesn't verify that we can send things to the wrong nodes when we don't have the data.

Status: Managed to write correct integration tests using one-shot LBPs to make sure I send queries to all nodes/shards.

### Wojciech's commits

What should I do with Wojciech's commits?
I'm considering squashing them (and introducing some changes to them so that resulting commit passes CI).
The changes are mostly mechanical and it would allow us to have atomic commits in this PR.

Status: Opened https://github.com/scylladb/scylla-rust-driver/pull/944 with his changes cleaned up


### Tablet feedback handled in Connection, not in Session

Tablet info is extracted from frame in Connection instead of Session + iterators.
I did look into moving that into session, but decided that it's too complicated and error-prone.
In the current implementation I can have certainty that we handle this for all requests.

Status: it will stay that way for now


### `FromCqlVal::from_cql` is tablets.rs

Is it possible for this to fail? I don't see how, but I'm not very familiar with this are of code.
If it can't fail, I could avoid introducing new error type.

Status: I added `.unwrap()`, it seems that this call can't fail.


### Parsing fail

What to do when parsing Tablet data from custom payload fails?
Erroring whole request seems excessive. Right now I just print a warning. Is it enough?

Status: Printing warning is enough, so I'll stick to that.

Fixes: #867 

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

avelanarius commented:
> @avelanarius suggested that there is "tablet_allocator_shuffle" HTTP api

[Based on discussion from Slack] We concluded that "tablet_allocator_shuffle" is great for manual testing, but for automatic testing there's not enough control to wait for it to take effect. Also, in practice it takes a long time (10s+) for tablet shuffle to actually happen.

avelanarius commented:
> Parsing fail
> What to do when parsing Tablet data from custom payload fails?
> Erroring whole request seems excessive. Right now I just print a warning. Is it enough?

I think it's enough.

piodul commented:
> ### Wojciech's commits
> 
> What should I do with Wojciech's commits? I'm considering squashing them (and introducing some changes to them so that resulting commit passes CI). The changes are mostly mechanical and it would allow us to have atomic commits in this PR.

I'd prefer the changes introduced by Wojciech's commits to be logically organized into commits. If the current organization of the code in the commits is bad (e.g. they weren't cleaned up fully and don't compile / pass tests / are illogical) then please put them into shape. If you want to preserve the credit then personally I think it is fine to add the original author as a co-author in the description.

> ### Integration tests
> 
> I wrote an integration test for tablets (last commit) but it doesn't always work, and I'm not sure how to write a proper test. The test first inserts some row. Each query is executed multiple times and the test checks that at least one execution was sent to wrong node and recevied tablet feedback. This is not always the case and I'm not sure how to force it.

Do you know why this happens? Won't such a query be sent to a random node/shard every time until it gets information about the partition's tablet?

> ### Tests with removed / decommisioned nodes
> 
> I'd like to test such scenarios, as suggested by @avelanarius , but I think we don't have a good way to do this now. There is proxy, but afaik it can only simulate network errors. It would be nice to use ccm (or something like this) in tests and have Rust API for it to manipulate cluster more easily. Need to figure out a way to test it now.

Yes, the rust driver repo doesn't have a good way to do tests with schema changes. I agree it would be nice to have it, but it is out of scope here and not a trivial effort.

> ### Parsing fail
> 
> What to do when parsing Tablet data from custom payload fails? Erroring whole request seems excessive. Right now I just print a warning. Is it enough?

Sending to the wrong node/shard is not an error. Warning sounds OK to me.

> ### Unknown replica
> 
> Is it possible to get a replica uuid in custom payload that we don't know about? What should we do in this case? Right now I'm just ignoring this replica, but I don't like this. This will cause load to be distributed less evenly because we skip one (or more) of replicas.

I think it is possible, for example if a node is added and some tablets are quickly moved to the new node before the driver refreshes information about topology. In such case it doesn't sound bad to ignore the replica (or maybe even send to a wrong node if we choose the unknown replica) until we refresh the metadata.

piodul commented:
As for this one:

> ### Tablet feedback handled in Connection, not in Session
> 
> Tablet info is extracted from frame in Connection instead of Session + iterators. I did look into moving that into session, but decided that it's too complicated and error-prone. In the current implementation I can have certainty that we handle this for all requests.

...not super happy, but I see why you decided to do it. I think that the execute and execute_iter deserve some refactoring and simplification. For now, I'm good with putting the sender into Conection.

Lorak-mmk commented on scylla/src/transport/iterator.rs:
```diff
@@ -272,20 +273,19 @@ impl RowIterator {
-                    )
-                } else {
-                    None
-                };
+            let replicas: Option<smallvec::SmallVec<[_; 8]>> = if let (Some(kstable), Some(token)) =
+                (statement_info.table.as_ref(), statement_info.token)
+            {
+                Some(
+                    config
+                        .cluster_data
+                        .get_token_endpoints_iter(kstable, token)
+                        .map(|(node, shard)| (node.clone(), shard))
+                        .collect(),
+                )
+            } else {
+                None
+            };
```
Formatting in a separate commit avoids this, but then we can't say that each commit passes static checks :/

Lorak-mmk replied:
Imo this is not an issue with this commit - this is an issue with tooling.
Diff in VSCode:
![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/e1f772b3-27c8-4301-ac4f-761be0f8c93e)

Diff in Sublime Merge:
![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/6ec7b9df-2437-4b5e-b8bc-30dc0b97d805)


I'd advise to look at diffs in tools other than GitHub - at least in cases like this

Lorak-mmk commented:
Reworked commit structure of Wojciechs commits and posted them as separate PR: https://github.com/scylladb/scylla-rust-driver/pull/944

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    /// 3. Tablets with unrecognized uuids in replica list are resolved again.
+    ///    If this is unsuccessful again then the tablet is removed.
+    ///    This can be skipped if we know we have no such tablets.
+    /// 4. Rarely the driver may need to re-create `Node` object for given node.
```
Rarely, the driver

wprzytula replied:
for a given node

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///    If this is unsuccessful again then the tablet is removed.
+    ///    This can be skipped if we know we have no such tablets.
+    /// 4. Rarely the driver may need to re-create `Node` object for given node.
+    ///    Old object is replaced with new in replica lists.
```
The old object is replaced with the new one

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///    This can be skipped if there were no re-created `Node` objects.
+    ///
+    /// In order to not perform unnecessary work during typical schema refresh
+    /// we avoid iterating trough tablets at all if steps 2-4 can be skipped.
```
through

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    /// * `removed_nodes`: Nodes that previously were present in ClusterData but are not anymore.
+    ///                    For any such node we should remove all tablets that have it in replica list.
+    ///                    This is because otherwise:
+    ///                    1. We would hold on to old `Node` objects, not allowing them to release memory.
```
We would hold/keep old

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///                    For any such node we should remove all tablets that have it in replica list.
+    ///                    This is because otherwise:
+    ///                    1. We would hold on to old `Node` objects, not allowing them to release memory.
+    ///                    2. Return removed nodes in LBP
```
We would return

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///                    This is because otherwise:
+    ///                    1. We would hold on to old `Node` objects, not allowing them to release memory.
+    ///                    2. Return removed nodes in LBP
+    ///                    3. When new node joins and becomes replica for this tablet, we would
+    ///                       not use it - instead we would keep querying a subset of replicas.
```
When a new node

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///                        then we would only query subset of replicas for the tablet,
+    ///                        potentially increasing load on those replicas.
+    ///                        The alternative is dropping the tablet immediately, but if there are a lot
+    ///                        of requests to range belonging to this tablet, then we would get a
```
requests to a range

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///                        The alternative is dropping the tablet immediately, but if there are a lot
+    ///                        of requests to range belonging to this tablet, then we would get a
+    ///                        lot of unnecessary feedbacks sent. Thus the current solution:
+    ///                        skipping unknown replicas and dropping the tablet if we still can't resolve
+    ///                        them after schema refresh.
```
schema refresh or topology refresh?

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -213,8 +339,75 @@ impl TabletsInfo {
+    ///
+    /// * `recreated_nodes`: There are some situations (IP change, DC / Rack change) where the driver
+    ///                      will create a new `Node` object for some node and drop the old one.
+    ///                      Tablet info would still contain old object, so the driver would not use
```
contain the old object

wprzytula commented on .github/workflows/rust.yml:
```diff
@@ -30,10 +30,14 @@ jobs:
       run: cargo clippy --verbose --all-targets
+    - name: Clippy check with all features
+      run: cargo clippy --verbose --all-targets --all-features
     - name: Cargo check without features
       run: cargo check --all-targets --manifest-path "scylla/Cargo.toml" --features ""
     - name: Cargo check with all serialization features
       run: cargo check --all-targets --manifest-path "scylla/Cargo.toml" --features "full-serialization"
+    - name: Cargo check with all features
+      run: cargo check --all-targets --manifest-path "scylla/Cargo.toml" --all-features
```
Not related to this PR: maybe we should also check `scylla-cql` separately with all features and with no features?
I can remember that at one moment I nearly rendered `scylla-cql` not compilable with no features, but with the workspace-level `cargo check` one couldn't experienced it, because `scylla` unconditionally enabled `chrono` dependency in `scylla-cql` IIRC.

Lorak-mmk replied:
Imho the `build` CI job should be split into static analysis and tests.
Then in static analysis one we could easily run a lot of combinations, because it should still take less time than tests

Lorak-mmk replied:
Opened https://github.com/scylladb/scylla-rust-driver/issues/982

wprzytula commented on Makefile:
```diff
@@ -25,10 +25,18 @@ check:
 check-without-features:
 	cargo check --manifest-path "scylla/Cargo.toml" --features "" --all-targets
 
+.PHONU: check-all-features
+check-all-features:
```
.PHONY

Lorak-mmk commented:
I've run some SCT job that uses cql-stress built with this branch.

![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/aa774858-a102-4d44-bd01-73ca81a72057)
![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/1c556ed7-9d14-45ff-8d16-b9ce9a98f709)


Initially metrics look like I expected them to do: a lot of non-token-aware queries initially after workload began (because driver had no tablet informations), but the amount quickly dropped to 0 as it fetched them.
After that node 10.4.11.143 was restarted (or at least that's what Grafana says). After the restart there were a lot of non-token-aware queries directed at this node and the amount seemed proportional to amount of all reads issued. There was no rapid decrease, unlike the beggining of the workload.
I don't understand this behavior :(

wprzytula commented:
> I've run some SCT job that uses cql-stress built with this branch.
> 
> ![image](https://private-user-images.githubusercontent.com/5958117/319382382-aa774858-a102-4d44-bd01-73ca81a72057.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyMTM2MzksIm5iZiI6MTcxMjIxMzMzOSwicGF0aCI6Ii81OTU4MTE3LzMxOTM4MjM4Mi1hYTc3NDg1OC1hMTAyLTRkNDQtYmQwMS03M2NhODFhNzIwNTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDRUMDY0ODU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmIxZTZkZTk3YmIxZmQ2OWRmMDQ1NGYxMmE4NzBmNGIxY2Y0ZmRkNGI4ODUxNGIzZTE4ZmFhY2FkODkxZmM0OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.la3NKblTjImB8X3k8VJ-gCrOyKHwGFqj-WHUEbtji-s) ![image](https://private-user-images.githubusercontent.com/5958117/319382418-1c556ed7-9d14-45ff-8d16-b9ce9a98f709.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyMTM2MzksIm5iZiI6MTcxMjIxMzMzOSwicGF0aCI6Ii81OTU4MTE3LzMxOTM4MjQxOC0xYzU1NmVkNy05ZDE0LTQ1ZmYtOGQxNi1iOWNlOWE5OGY3MDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MDRUMDY0ODU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzAzNDcxNTA2OGUwMmU4MDM0OTUxNDBhYTM3YzhiMGFlZDk2NTY2MTEwNjA1NDc3YWY2MDVmOTM5NWUwZTY0NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4RBo8GtZHCP8HsxrBSDskpRXaLuuhm6YPXb6DjyDfi4)
> 
> Initially metrics look like I expected them to do: a lot of non-token-aware queries after workload began (because driver had no tablet informations), but the amount quickly dropped to 0 as it fetched them. After that node 10.4.11.143 was restarted (or at least that's what Grafana says). After the restart there were a lot of non-token-aware queries directed at this node and the amount seemed proportional to amount of all reads issued. There was no rapid decrease, unlike the beggining of the workload. I don't understand this behavior :(

Any insight from the logs?

Lorak-mmk commented:
Didn't look at them because I went to sleep. I'll run the test again tomorrow to investigate more.
Also, when I changed the formula for "Non-Token Aware Queries" to `sum(rate(scylla_storage_proxy_coordinator_reads_coordinator_outside_replica_set[1m])) + sum(rate(scylla_storage_proxy_coordinator_writes_coordinator_outside_replica_set[1m]))` there seem to not be any non-token aware queries.
Which is unexpected - there should be some in the beggining when driver doesn't have any tablet information. Maybe the test somehow didn't use tablets despite being run with latest build of Scylla?

The formula used for "Non-Token Aware Queries" on screenshot is:
```scalar(sum(rate(scylla_storage_proxy_coordinator_cas_total_operations{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]|$^"}[1m])) <=bool 0)*scalar(sum(rate(scylla_storage_proxy_replica_received_counter_updates{cluster=~"$cluster|$^"}[1m]))<=bool 0) *clamp_min(sum(rate(scylla_cql_reads{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) - sum(rate(scylla_storage_proxy_coordinator_reads_local_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) + sum(rate(scylla_cql_inserts{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m]) + rate(scylla_cql_updates{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m]) + rate(scylla_cql_deletes{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) - sum(rate(scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]),0)
```
given to me by @michoecho because the default formula contains a bug and doesn't work. @michoecho suspects that the jump visible in the screenshot is in fact some bug in monitoring, not the jump that I expected from tablet-aware driver.
That would be consistent with the theory that somehow tablets are not used. I'll need to investigate more.

michoecho commented:
> ```
> scalar(sum(rate(scylla_storage_proxy_coordinator_cas_total_operations{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]|$^"}[1m])) <=bool 0)*scalar(sum(rate(scylla_storage_proxy_replica_received_counter_updates{cluster=~"$cluster|$^"}[1m]))<=bool 0) *clamp_min(sum(rate(scylla_cql_reads{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) - sum(rate(scylla_storage_proxy_coordinator_reads_local_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) + sum(rate(scylla_cql_inserts{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m]) + rate(scylla_cql_updates{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m]) + rate(scylla_cql_deletes{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) - sum(rate(scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]),0)
> ```
> 
> 
>     
>   
> 
> given to me by @michoecho because the default formula contains a bug and doesn't work

That's the old (wrong) formula. The new one is
```
clamp_min(sum(rate(scylla_cql_reads{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) - sum(rate(scylla_storage_proxy_coordinator_reads_local_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) + sum(rate(scylla_cql_inserts{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m]) + rate(scylla_cql_updates{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m]) + rate(scylla_cql_deletes{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]) - sum(rate(scylla_storage_proxy_coordinator_total_write_attempts_local_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]]),0)
```

(It doesn't have the `scalar` factors).

> @michoecho suspects that the jump visible in the screenshot is in fact some bug in monitoring

Rationale: the formula uses `rate(scylla_cql_reads[1m]) - rate(scylla_storage_proxy_coordinator_reads_local_node[1m])`.
This apparently gives a result of several thousands per second. (The "jump").

But if you just plot `scylla_cql_reads - scylla_storage_proxy_coordinator_reads_local_node`, you will observe that the result is actually constant at the same time.

Which means that Prometheus has two counters which grow by the same amount in the same time interval, but somehow it calculates different a `rate` for them. Which is unexpected.

If you do the `rate` calculation manually, it seems that Prometheus underestimates the for `scylla_storage_proxy_coordinator_total_write_attempts_local_node` for one scrape period. And that's the source of the initial *jump*.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -95,44 +96,120 @@ pub(crate) struct Tablet {
 #[derive(Clone, Debug, Default)]
 pub(crate) struct TableTablets {
     tablet_list: Vec<Tablet>,
+    /// In order to make typical tablet maintance faster
+    /// we remember if there were any tablets that have unrecognized uuids in replica list.
+    /// If there were none, and few other conditions are satisfied, we can skip nearly whole maintanace.
+    /// This flag may be falsely true: if we add tablet with unknown replica but later
+    /// overwrite it with some other tablet.
```
If you know a simpler way please describe it, otherwise I'm not sure what is this comment supposed to achieve

wprzytula replied:
This was just an initial impression that I had about the logic and its explanation, and I wanted to share it as a feedback. I don't have specific tips for making it clearer yet, though,

Lorak-mmk commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,628 @@
+        );
+    }
+
+    let selected_tablets_rows = session.query(
+        "select last_token, replicas from system.tablets WHERE keyspace_name = ? and table_name = ? ALLOW FILTERING",
+        &(ks.as_str(), table.as_str())).await.unwrap().rows.unwrap();
```
I highly doubt we'll encounter such problems in tests. If we do at some point we can do a paged query then.

wprzytula replied:
@Lorak-mmk bump

Lorak-mmk replied:
I marked it as resolved because I changed it to `query_iter` locally

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+
+        custom_payload.insert(CUSTOM_PAYLOAD_TABLETS_V1_KEY.to_string(), data);
+
+        assert!(matches!(
+            RawTablet::from_custom_payload(&custom_payload),
+            Some(Err(TabletParsingError::Parse(_)))
+        ));
```
Replaced both occurrences in this file

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -95,44 +96,120 @@ pub(crate) struct Tablet {
+                *node = Arc::clone(new_node);
+            }
+
+            if any_updated {
+                // Now that we know we have some nodes to update we need to go over
+                // per-dc nodes and update them too.
+                for (_, dc_nodes) in self.replicas.per_dc.iter_mut() {
+                    for (node, _) in dc_nodes.iter_mut() {
+                        if let Some(new_node) = recreated_nodes.get(&node.host_id) {
+                            *node = Arc::clone(new_node);
+                        }
+                    }
+                }
+            }
```
Opps, of course it should - nice catch

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -95,44 +96,120 @@ pub(crate) struct Tablet {
+                    true
+                }
+                Err((_, failed)) => {
+                    warn!("Nodes ({}) listed as replicas for a tablet are not present in ClusterData.known_peers, \
+                           despite topology refresh. Removing problematic tablet.", failed.iter().format(", "));
```
How? Print it's token range?

wprzytula replied:
If the token range serves as the best identifier of a tablet, then yes. Anything that could come in handy while debugging will be appreciated.

Lorak-mmk replied:
The token range alone is not precise - keyspace name and table name are missing, but we don't have access to them in `Tablet`

Lorak-mmk replied:
Token range is not precise. `Table` method that this log is located in could return Result with list of failed replicas instead of bool. Then I'd have to print the token range and table name in `TableTablets` - which doesn't store table name, but I can add it - it's one allocation per table, so it's ok

wprzytula replied:
Yes, one allocation per table seems acceptable.

Lorak-mmk replied:
Moved the error message to `TableTablets::perform_maintanance`, with precise tablet identification

Lorak-mmk replied:
I also moved the message from `Tablet::from_raw_tablet` to cluster.rs and improved in in the same way (printing precise tablet info)

Lorak-mmk commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,628 @@
+        }
+    }
+
+    assert!(present_tablets.iter().all(|x| *x));
```
The function is trying to find a key per tablet. It tries 1000 keys. Given the initial amount of tablets (16) set in the tests it should easily find those keys. The assert verifies that it in fact did.

wprzytula replied:
Please write it clearly in comments there.

Lorak-mmk commented on scylla/src/utils/test_utils.rs:
```diff
@@ -93,6 +92,42 @@ pub fn create_new_session_builder() -> GenericSessionBuilder<impl SessionBuilder
+    let result = session
+        .query(
+            "SELECT * FROM system_schema.scylla_keyspaces WHERE keyspace_name=?",
+            [keyspace].as_ref(),
```
I just removed the whole function, it was not used by any test anyway.

Lorak-mmk commented on scylla/src/transport/cluster.rs:
```diff
@@ -517,6 +537,17 @@ impl ClusterWorker {
+                        // If the channel was closed then cluster was dropped, we can stop working
+                        return;
+                    }
+                    let mut new_cluster_data: ClusterData = self.cluster_data.load().as_ref().clone();
```
Afaik `load()` is preferable when we don't hold the object for a long time according to https://docs.rs/arc-swap/latest/arc_swap/struct.ArcSwapAny.html#method.load

wprzytula replied:
But for instance:
```rust
fn change_node_down_marker(&mut self, addr: SocketAddr, is_down: bool) {
    let cluster_data = self.cluster_data.load_full();
    // We need to iterate through the whole map here, but there will rarely more than ~100 nodes,
    // and changes of down marker are infrequent enough to afford this. As an important tradeoff,
    // we only keep one hashmap of known_peers, which is indexed by host IDs for node identification.
    let node = match cluster_data
        .known_peers
        .values()
        .find(|&peer| peer.address == NodeAddr::Translatable(addr))
    {
        Some(node) => node,
        None => {
            warn!("Unknown node address {}", addr);
            return;
        }
    };

    node.change_down_marker(is_down);
}
```
uses `load_full` even though it doesn't seem to hold the `cluster_data` for a long time.

Lorak-mmk replied:
Are you asking me why the author of this code wrote it this way? You should ask the author, not me. I can only explain why I wrote my code the way I wrote it.
And I wrote my code this way because the `arc_swap` documentation says I should.

wprzytula replied:
I'm asking what are your thoughts after seeing this code. How do you explain the discrepancy? Do you believe that the code I gave uses `load_full` unnecessarily?

Lorak-mmk replied:
Idk, maybe - you can open an issue about it if you want, as it's not related to this PR.

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,628 @@
+        Some((&self.target.0, self.target.1))
+    }
+
+    fn fallback<'a>(
+        &'a self,
+        _query: &'a RoutingInfo,
+        _cluster: &'a ClusterData,
+    ) -> FallbackPlan<'a> {
+        Box::new(std::iter::once((&self.target.0, self.target.1)))
+    }
```
It is nowhere stated that `fallback` has to repeat the target that was returned by `pick`, as well as that `fallback` is "a full policy". Perhaps we should settle on one leading interpretation and document it clearly?

Lorak-mmk replied:
I made fallback return empty list

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        self.tablet_list.drain(left_idx..right_idx);
+        self.tablet_list.insert(left_idx, tablet);
+    }
+}
```
I don't exactly remember - I think @avelanarius said something about a maximum of 100k per table.
We could use some tree-like structure instead of Vec for faster insertion / deletion, possibly at the cost of slower lookup / increased memory usage.
Maybe it's possible to write this with `BTreeMap` (if not we'd need to use external crate / hand written data structure), but first we'd need some benchmarks to see if it is really faster.

Fortunately the implementation doesn't leak outside this module so we can easily change this later.

Lorak-mmk replied:
This PR is already quite big, so I think it's best to postpone evaluating different implementation. When we merge this I'll open an issue about benchmarking tree-based implementation.

Lorak-mmk replied:
Opened: https://github.com/scylladb/scylla-rust-driver/issues/983

Lorak-mmk commented on scylla/src/transport/cluster.rs:
```diff
@@ -508,7 +544,14 @@ impl ClusterData {
 
         for (table, raw_tablet) in raw_tablets.into_iter() {
-            let tablet = Tablet::from_raw_tablet(&raw_tablet, replica_translator);
+            // Should we skip tablets that belong to keyspace not present in
+            // self.keyspaces? The keyspace could have been, without drivers knowledge:
+            // 1. Dropped - in which case we'll remove it's info soon (when refreshing
+            // topology) anyway.
+            // 2. Created - no harm in storing the info now.
```
Because we update metadata periodically (and on topology change), so it's possible we didn't fetch it yet.

Lorak-mmk commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,628 @@
+                }
+            }
+
+            // There is no special meaning to "3", I just thought it gives low
+            // enough chances of failure.
+            assert!(statements_with_feedback > (value_lists.len() / 3));
```
No specific reason and no calculations, just intuition

Lorak-mmk replied:
Hm... I read my comment above now and I have no idea where I got "30" from.
But other change (when there is no tablet in local data then return empty replica list, not vnode replica list) should make this test totally non-flaky, even when expecting to receive feedback for all tablets, not a subset of them.

Lorak-mmk replied:
I've run the test a 100 times after changes (requiring that always feedbacks > 0), so I can merge it with `test_tablet_awareness_works_with_full_info` because the workaround with one-shot LBPs is no longer necessary,

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct RawTablet {
+    first_token: Token,
+    last_token: Token,
```
Imho "inclusive" is the default - as the name suggests. Here both values are inclusive, so I didn't see the need to specify it in the name

wprzytula replied:
OTOH, it's a predominant convention to represent ranges as [start, end), so especially when using another convention, being explicit is encouraged.

Lorak-mmk replied:
I added doc comments to not make the names longer and more verbose. Is that enough?

Lorak-mmk replied:
@wprzytula Current struct definition fyi:
```rust
#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct Tablet {
    /// First token belonging to the tablet, inclusive
    first_token: Token,
    /// Last token belonging to the tablet, inclusive
    last_token: Token,
    replicas: TabletReplicas,
    /// If any of the replicas failed to resolve to a Node,
    /// then this field will contain the original list of replicas.
    failed: Option<RawTabletReplicas>,
}
```

Lorak-mmk commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1316,7 +1364,7 @@ mod tests {
                 routing_info: RoutingInfo {
                     token: Some(Token::new(160)),
-                    keyspace: Some(KEYSPACE_NTS_RF_2),
+                    table: Some(table_specs.get("KEYSPACE_NTS_RF_2").unwrap()),
```
Can you explain what does this change improve? I don't see an advantage

wprzytula replied:
> Strings are much more error-prone.

If one changes the content of the constant in a distant module - locator - the test will be broken. This is not how it should work; it should be that the exact content of a constant is up to be changed, and it's the name of the constant that is used everywhere else in the code.

Lorak-mmk replied:
Now that `TableSpec` doesn't need to allocate this can actually be made const global again.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        ]);
+
+        SerializeCql::serialize(&value, &col_type, CellWriter::new(&mut data)).unwrap();
+        println!("{:?}", data);
```
Changed to `tracing::debug!`

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+            } else {
+                // TODO: Should this be an error? When can this happen?
+                warn!("Node {replica} from system.tablets not present in ClusterData.known_peers. Skipping this replica");
+                None
```
This is explained and handled in a commit that adds maintenance procedure for tablets.

Lorak-mmk commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -82,13 +82,37 @@ impl ReplicaLocator {
-        _table: &TableSpec,
+        table: &TableSpec,
     ) -> ReplicaSet<'a> {
+        if let Some(tablets) = self.tablets.tablets_for_table(table) {
+            let replicas: Option<&[(Arc<Node>, u32)]> = if let Some(datacenter) = datacenter {
+                tablets.dc_replicas_for_token(token, datacenter)
+            } else {
+                tablets.replicas_for_token(token)
+            };
```
Let's do this when we remove "unstable-tablets" feature, right now tablets are actually an abnormal path

Lorak-mmk commented on scylla/src/transport/load_balancing/mod.rs:
```diff
@@ -24,7 +27,7 @@ pub struct RoutingInfo<'a> {
 
     /// Information about token and keyspace is the basis of token-aware routing.
     pub token: Option<Token>,
-    pub keyspace: Option<&'a str>,
+    pub table: Option<&'a TableSpec>,
```
How would you like the docstring changed (apart from changing `keyspace` to `keyspace and table`? It looks to me like it's still accurate - it only says about token-aware routing, it doesn't say if token awareness is done trough vnodes or tablets

wprzytula replied:
```rust
/// Information that are the basis of token-aware routing:
/// - token and keyspace for vnodes-based routing;
/// - token, keyspace and table for tablets-based routing. 
```

Lorak-mmk commented on scylla/src/transport/cluster.rs:
```diff
@@ -517,6 +537,17 @@ impl ClusterWorker {
                         None => return, // If refresh_channel was closed then cluster was dropped, we can stop working
                     }
                 }
+                tablets_count = self.tablets_channel.recv_many(&mut tablets, 8192) => {
```
Justification: it's the size of the channel. I want to be able to receive everything in the channel.

I changed it to a constant, so it should be clear now

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        );
+    }
+
+    #[test]
+    fn replace_multiple_tablets_middle() {
+        let mut tablets = Default::default();
+        insert_ranges(
+            &mut tablets,
+            &[
+                (-2000000, -1800001),
+                (-1800000, -1600001),
+                (-1600000, -1400001),
+                (-1400000, -1200001),
+                (-1200000, -1000000),
+            ],
+        );
+
+        // Replacing 3 middle tablets
+        insert_ranges(&mut tablets, &[(-1750000, -1250000)]);
+        verify_ranges(
+            &tablets,
+            &[
+                (-2000000, -1800001),
+                (-1750000, -1250000),
+                (-1200000, -1000000),
+            ],
+        );
+    }
+}
```
I added the following comment to `TableTablets` struct:
```
/// Container for tablets of a single table.
/// 
/// It can be viewed as a set of non-overlapping Tablet objects.
/// It has 2 basic operations:
/// 1. Find a tablet for given Token
/// 2. Add a new tablet.
/// 
/// Adding new Tablet will first remove all tablets that overlap with the new tablet.
```

is this explanation enough?

Lorak-mmk commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,628 @@
+    statement: &PreparedStatement,
+    values: impl SerializeRow,
+) -> Result<Vec<QueryResult>, QueryError> {
+    let mut tasks = vec![];
+    for node in cluster.get_nodes_info() {
+        let shard_count: u16 = node.sharder().unwrap().nr_shards.into();
+        for shard in 0..shard_count {
+            let mut stmt = statement.clone();
+            let values_ref = &values;
+            let policy = SingleTargetLBP {
+                target: (node.clone(), Some(shard as u32)),
+            };
+            let execution_profile = ExecutionProfile::builder()
+                .load_balancing_policy(Arc::new(policy))
+                .build();
+            stmt.set_execution_profile_handle(Some(execution_profile.into_handle()));
+
+            let task = async move { session.execute(&stmt, values_ref).await };
+
+            tasks.push(task);
+        }
+    }
+
+    try_join_all(tasks).await
```
try_join_all will allocate it anyway, but sure, here the code is still not very hard to understand, so I'll apply the suggestion.

wprzytula commented on scylla/src/transport/cluster.rs:
```diff
@@ -351,6 +353,40 @@ impl ClusterData {
 
+        #[cfg(feature = "unstable-tablets")]
+        {
+            let mut removed_nodes = HashSet::new();
+            for old_peer in known_peers {
+                if !new_known_peers.contains_key(old_peer.0) {
+                    removed_nodes.insert(*old_peer.0);
+                }
+            }
```
If you really don't like FP in this case (although it's more readable to me), consider another way of reducing scope of mutability:
```suggestion
            let removed_nodes = {
                let mut removed_nodes = HashSet::new();
                for old_peer in known_peers {
                    if !new_known_peers.contains_key(old_peer.0) {
                        removed_nodes.insert(*old_peer.0);
                    }
                }
                removed_nodes
            }
```

Lorak-mmk replied:
I like this new suggestion, thanks

Lorak-mmk commented on scylla/src/transport/cluster.rs:
```diff
@@ -351,6 +353,40 @@ impl ClusterData {
+                }
+            };
+
+            let mut recreated_nodes = HashMap::new();
+            for (old_peer_id, old_peer_node) in known_peers {
+                if let Some(new_peer_node) = new_known_peers.get(old_peer_id) {
+                    if !Arc::ptr_eq(old_peer_node, new_peer_node) {
+                        recreated_nodes.insert(*old_peer_id, Arc::clone(new_peer_node));
+                    }
+                }
+            }
```
Applied the suggestion from the comment above here as well

Lorak-mmk commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,628 @@
+        .collect::<Vec<_>>();
+    selected_tablets.sort_unstable_by(|a, b| a.last_token.cmp(&b.last_token));
+
+    let mut tablets = Vec::new();
+    let mut first_token = i64::MIN;
+    for tablet in selected_tablets {
```
There will still be `mut first_token` - it will be more confusing if we are using functional APIs that mutate some outside variable.
But it actually can be done with `fold` - I think it is one of the rare cases where it is a proper tool.
There is still `tablets.push`, but there are not `mut` vars outside fold call.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        // to `deser_cql_value`.
+        let result: RawTabletPayload = FromCqlVal::from_cql(cql_value).unwrap();
+
+        let replicas = match result
+            .2
+            .into_iter()
+            .map(|(uuid, shard_num)| match shard_num.try_into() {
+                Ok(s) => Ok((uuid, s)),
+                Err(_) => Err(shard_num),
+            })
+            .collect::<Result<Vec<(Uuid, u32)>, _>>()
+        {
+            Ok(r) => r,
+            Err(shard_num) => return Some(Err(TabletParsingError::ShardNum(shard_num))),
+        };
```
I remember reading about optimization in `Vec` that prevents reallocation in case where result and source have the same size - but I'm not sure it will be applied there (and it is an optimization, so it is not guaranteed).

Lorak-mmk replied:
Vec docs mention this optimization (with a note that it is not guaranteed to be stable, of course): https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-FromIterator%3CT%3E-for-Vec%3CT%3E
Not sure if this will be applied here as we use Result, but maybe it will?

Lorak-mmk replied:
Anyway, I don't see any way to avoid this vector now, so marking it as resolved.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        let mut data = vec![];
+
+        let value = CqlValue::Tuple(vec![
+            Some(CqlValue::BigInt(1234)),
+            Some(CqlValue::BigInt(2137)),
+            Some(CqlValue::List(vec![
+                CqlValue::Tuple(vec![
+                    Some(CqlValue::Uuid(Uuid::from_u64_pair(1, 2))),
+                    Some(CqlValue::Int(15)),
+                ]),
+                CqlValue::Tuple(vec![
+                    Some(CqlValue::Uuid(Uuid::from_u64_pair(3, 4))),
+                    Some(CqlValue::Int(19)),
+                ]),
+            ])),
+        ]);
```
That makes a lot of sense, done

muzarski commented on .github/workflows/tablets.yml:
```diff
@@ -0,0 +1,38 @@
+        run: SCYLLA_URI=127.0.1.1:9042 SCYLLA_URI2=127.0.1.2:9042 SCYLLA_URI3=127.0.1.3:9042 cargo test --verbose --features unstable-tablets
+
+      - name: Remove tablets cluster
+        run: ccm remove tablets
```
nit: missing newline at the end of file

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        }
+    }
+
+    #[test]
+    fn table_tablets_single() {
+        let mut tablets = Default::default();
+
+        insert_ranges(&mut tablets, &[(-200, 1000)]);
+        verify_ranges(&tablets, &[(-200, 1000)]);
+
+        assert_eq!(
+            tablets.tablet_for_token(Token::new(-1)),
+            Some(&tablets.tablet_list[0])
+        );
+        assert_eq!(
+            tablets.tablet_for_token(Token::new(0)),
+            Some(&tablets.tablet_list[0])
+        );
+        assert_eq!(
+            tablets.tablet_for_token(Token::new(1)),
+            Some(&tablets.tablet_list[0])
+        );
+        assert_eq!(
+            tablets.tablet_for_token(Token::new(-200)),
+            Some(&tablets.tablet_list[0])
+        );
+        assert_eq!(tablets.tablet_for_token(Token::new(-201)), None);
+        assert_eq!(
+            tablets.tablet_for_token(Token::new(1000)),
+            Some(&tablets.tablet_list[0])
+        );
+        assert_eq!(tablets.tablet_for_token(Token::new(1001)), None);
+    }
```
I'm not sure what would you want to make constant and how it improves this specific test. Could you elaborate?

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+        assert_eq!(
+            tablet,
+            RawTablet {
+                first_token: Token::new(1235),
+                last_token: Token::new(2137),
```
2137 is no longer there, so this I'll skip this

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+
+    #[test]
+    fn test_raw_tablet_deser_trash() {
+        let mut custom_payload = HashMap::new();
+        custom_payload.insert(CUSTOM_PAYLOAD_TABLETS_V1_KEY.to_string(), vec![1, 2, 3]);
```
This function is so small that I don't see the advantage of that.

muzarski replied:
> Using iterators + `collect()` to create a `HashMap` instead of `insert`s seems much cleaner to me, as it avoids mutability.

Why not just:
```rust
let custom_payload =
            HashMap::from([(CUSTOM_PAYLOAD_TABLETS_V1_KEY.to_string(), vec![1, 2, 3])]);
```

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,630 @@
+            PeerEndpoint {
+                host_id: Uuid::from_u64_pair(1, 1),
+                address: addr,
+                datacenter: Some("dc1".to_string()),
```
If you mistakenly change `"dc1"` (a random keyboard typo) to `"dc1l"`, the compiler won't complain. If you however do that with a constant: `DC1` -> `DC1l`, then the complainer will shout at you (undefined item).

My rule of thumb is that if in any two spots in code I expect **the same** value, then I'm in favor of turning it into a constant, to explicitly show this relationship. This speeds up comprehending the relationships there.

Lorak-mmk commented on scylla/Cargo.toml:
```diff
@@ -53,6 +53,7 @@ url = { version = "2.3.1", optional = true }
 base64 = { version = "0.21.1", optional = true }
 rand_pcg = "0.3.1"
 socket2 = { version = "0.5.3", features = ["all"] }
+lazy_static = "1"
```
In current local version of PR I have added an explanation in the commit message as requested by @wprzytula . Do you think splitting is necessary or explanation in commit message is enough?

muzarski replied:
I think the explanation is enough.

muzarski commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -208,6 +209,14 @@ impl PreparedStatement {
+    pub fn get_table_spec(&self) -> Option<&TableSpec> {
+        self.get_prepared_metadata()
+            .col_specs
+            .first()
```
Oh, you are right, I didn't notice the other methods that make use of the same assumption.

However, I still think that we should investigate if scylla can provide different per-col specs. But it's not directly related to this PR, so as of now, this method looks good to me.

wprzytula replied:
We've already investigated that: https://github.com/scylladb/scylladb/issues/17788#issuecomment-1995992185

Lorak-mmk replied:
I think we can close this for now

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -217,6 +217,8 @@ pub(crate) struct QueryResponse {
     pub(crate) tracing_id: Option<Uuid>,
     pub(crate) warnings: Vec<String>,
+    #[allow(dead_code)] // This is not exposed to user (yet?)
+    pub(crate) custom_payload: Option<HashMap<String, Vec<u8>>>,
```
It looks like it would still require more changes in scylla-cql than I'm comfortable with in this PR.
custom_payload is rarely used feature (pre-tablets), so we can assume that without tablets those allocations won't happen.
With tablets it's 1 additional allocation. I think we can live with this for now.

I'll open an issue so we can improve that in the future.

Lorak-mmk replied:
Opened: https://github.com/scylladb/scylla-rust-driver/issues/984

Lorak-mmk commented on scylla-cql/src/frame/mod.rs:
```diff
@@ -204,20 +205,22 @@ pub fn parse_response_body_extensions(
-        // For now, just skip it
+    let custom_payload = if flags & FLAG_CUSTOM_PAYLOAD != 0 {
         let body_len = body.len();
         let buf = &mut &*body;
-        types::read_bytes_map(buf)?;
+        let payload_map = types::read_bytes_map(buf)?;
         let buf_len = buf.len();
         body.advance(body_len - buf_len);
-    }
+        Some(payload_map)
+    } else {
+        None
+    };
```
Opened: https://github.com/scylladb/scylla-rust-driver/issues/984

Lorak-mmk commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -122,7 +125,8 @@ or refrain from preferring datacenters (which may ban all other datacenters, if
             StatementType::NonLwt
         };
-        if let Some(ts) = &routing_info.token_with_strategy {
+        if let (Some(ts), Some(kstable)) = (&routing_info.token_with_strategy, query.table.as_ref())
```
I fixed the occurences you pointed out and some other that I found myself. It's interesting that clippy doesn't warn about this, do you know why?

muzarski replied:
I have no idea, unfortunately. I even checked a very simple scenario:
```rust
#[derive(Debug)]
struct Foo;

impl AsRef<Foo> for Foo {
    fn as_ref(&self) -> &Foo {
        self
    }
}

fn foo(f: &Foo) {
    dbg!(f);
}

fn main() {
    let f = Foo;
    let fref: &&Foo = &f.as_ref();
    foo(fref);
}
```

Clippy does not complain here as well...

wprzytula commented:
> Status: Printing warning is enough, so I'll stick to that.

I think printing using `tracing::error!()` is a better option, as failing to parse tablet payload is definitely an error.

wprzytula commented on scylla/src/transport/iterator.rs:
```diff
@@ -248,11 +248,12 @@ impl RowIterator {
             };
 
+            let kstable = config.prepared.get_table_spec();
             let statement_info = RoutingInfo {
                 consistency,
                 serial_consistency,
                 token,
-                keyspace: config.prepared.get_keyspace_name(),
+                table: kstable,
```
OK, then choose the convention you find the most reasonable and stick to it.

Lorak-mmk replied:
Renamed `TableSpec` arguments to `table_spec` - `table` / `keyspace` are used for strings. One exception: `RoutingInfo` field is still called `table`, I think that changing it would harm readability so I decided against it.

Lorak-mmk commented:
v2:
- Merged 2 of the integration tests, because the workaround that forced me to split them into two initially is no longer necessary
- Addressed all the review comments - that was quite a lot of work...
- Removed unstable-tablets feature - no need to keep tablets behind feature flag because they don't modify public API

I still need to add at least basic unit tests for maintenance procedures, but this isn't a stopper for review.

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,901 @@
+    #[test]
+    fn test_adding_tablets_non_overlapping() {
+        let mut tablets = TableTablets::new_for_test();
+        const RANGES: [(i64, i64); 5] = [
+            (-2000000, -1900001),
+            (-1900000, -1700001),
+            (-1700000, -1),
+            (0, 19),
+            (20, 10000),
+        ];
+
+        insert_ranges(&mut tablets, RANGES.as_slice());
+        verify_ranges(&tablets, RANGES.as_slice());
```
Why not make `RANGES` a slice in the first place?

```suggestion
        const RANGES: &[(i64, i64)] = &[
            (-2000000, -1900001),
            (-1900000, -1700001),
            (-1700000, -1),
            (0, 19),
            (20, 10000),
        ];

        insert_ranges(&mut tablets, RANGES);
        verify_ranges(&tablets, RANGES);
```

wprzytula commented on scylla/src/transport/locator/test.rs:
```diff
@@ -245,6 +257,7 @@ fn test_simple_strategy_replicas(locator: &ReplicaLocator) {
                 replication_factor: 3,
             },
             None,
+            &TableSpec::borrowed("ks", "invalid"),
```
Please make this a named constant and use in all places.

wprzytula replied:
What is the intention here? I guess that the table is intended as invalid, what about keyspace? `"ks"` tells me nothing.

Lorak-mmk replied:
Constant was already declared actually. Changed those `borrowed` calls to use the constant. Changed the constant to `TableSpec::borrowed("invalid", "invalid")`

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,586 @@
+        .unwrap();
+}
+
+/// Tests that, when using DefaultPolicy with TokenAwareness and querying table
+/// that uses tablets:
+/// 1. When querying data that belongs to tablet we didn't receive yet we will
+///    receive this tablet data at some point.
+/// 2. When we have all tablets info locally then we'll never receive tablet info.
+///
+/// The test first sends 100 queries per tablet and expects to receive tablet info.
+/// After that we know we have all the info. The test sends the statements again
+/// and expects to not receive any tablet info.
```
You seem to use both "tablet info" and "tablet data" terms. Are they the same?

Lorak-mmk replied:
Yes, I'll stick to info (or just tablet)

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,586 @@
+    }
+}
+
+/// Verifies that LWT optimization (sending LWT queries to a single replica always)
```
I'm unsure about this definition of the LWT optimization. I believe it could be better described by _sending LWT queries to the replicas in a fixed order_; there can be more replicas in the query plan than just a single one.

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,586 @@
+                )]));
+            }
+
+            // Unlike test_tablet_awareness_works_with_full_info I use "send_statement_everywhere",
+            // in order to make the test faster (it sends request per shard, not 100 requests).
```
sends one/a request per shard

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -36,7 +36,7 @@ pub struct SchemaChange {
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct TableSpec<'a> {
     pub ks_name: Cow<'a, str>,
     pub table_name: Cow<'a, str>,
```
One more argument against deriving everything: it will increase code size

Lorak-mmk replied:
+ please let's keep the PR comments for discussion about this PR. Open issue if you want to discuss another matter.

Lorak-mmk commented on scylla/src/transport/cluster.rs:
```diff
@@ -409,24 +458,30 @@ impl ClusterData {
-        self.get_token_endpoints_iter(keyspace, token)
+    pub fn get_token_endpoints(
+        &self,
+        keyspace: &str,
+        table: &str,
```
It's not related to this PR

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -116,6 +117,40 @@ pub enum CqlValue {
+        TableSpec::owned(self.ks_name.into_owned(), self.table_name.into_owned())
+    }
+
+    pub fn to_owned(&self) -> TableSpec<'static> {
+        TableSpec::owned(self.ks_name().to_owned(), self.table_name().to_owned())
+    }
```
I don't see how this trait could be implemented here. Implementation like this:
```rust
impl<'a> ToOwned for TableSpec<'a> {
    type Owned = TableSpec<'static>;

    fn to_owned(&self) -> Self::Owned {
        todo!()
    }
}
```
has a conflict:
```
error[E0119]: conflicting implementations of trait `ToOwned` for type `TableSpec<'_>`
   --> scylla-cql/src/frame/response/result.rs:145:1
    |
145 | impl<'a> ToOwned for TableSpec<'a> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `alloc`:
            - impl<T> ToOwned for T
              where T: Clone;
```

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -36,7 +36,7 @@ pub struct SchemaChange {
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct TableSpec<'a> {
     pub ks_name: Cow<'a, str>,
     pub table_name: Cow<'a, str>,
```
I don't think they need to be `pub(crate)`, they can be private.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,901 @@
+        }
+    }
+
+    pub(crate) fn range(&self) -> (Token, Token) {
+        (self.first_token, self.last_token)
+    }
```
This is not a range meant for iteration, just for printing a debug info. I don't see how this change would improve the code, but it makes it more complicated.

wprzytula replied:
`Range` object is not only intended for iteration, but it also provides `start()` and `end()` methods, too. For me it would make a great sense to return `Range` object from `range` method.
But OK, it's a nit.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,901 @@
+        table_spec: &'b TableSpec<'b>,
+    ) -> Option<&'a TableTablets> {
+        #[derive(Hash)]
+        struct TableSpecQueryKey<'a> {
+            table_spec: &'a TableSpec<'a>,
+        }
```
I'm not sure what you mean, and quick googling also didn't help. What optimization will be appliced for a single element tuple struct that won't be for a single element normal struct?

wprzytula replied:
I was convinced that only structs with a single unnamed field are subject to the newtype optimisation, i.e. a guarantee that their inner representation is the same as the representation of their field.
I can't find any sources though, so it's probably not true.

muzarski commented on scylla/src/transport/cluster.rs:
```diff
@@ -491,7 +532,22 @@ impl ClusterData {
+            // topology) anyway.
+            // 2. Created - no harm in storing the info now.
+            //
+            // So I think we can safely skip checking keyspace presence.
```
If these are the only cases (and I can't think of any other myself) when we get tablet info about keyspace not present in `self.keyspaces`, then I agree.

muzarski commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -114,39 +118,116 @@ pub(crate) struct Tablet {
+    fn re_resolve_replicas(
+        &mut self,
+        replica_translator: impl Fn(Uuid) -> Option<Arc<Node>>,
+    ) -> Result<(), Vec<Uuid>> {
```
I think the comment is outdated, we no longer return `bool`.

Lorak-mmk commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,681 @@
+                .map(|x| x.as_slice())
+                .unwrap_or(&[])
+        })
+    }
```
This has no practical implications - in locator we'll do `.unwrap_or(&[])` anyway.
It is not possible for `tablet.replicas.per_dc.get(dc)` to return `Some(&[])`. `per_dc` keys are only inserted with at least one value.
That means your solution doesn't differentiate between no tablet at all and no replica for given dc while mine does - I think those 2 scenarios are different enough that we should differentiate them.

Mine solution doesn't differentiate between no replica for given dc and empty list of replicas for given dc (which is impossible) - imo it's not a problem.

Lorak-mmk commented:
- Added maintenance procedure tests
- Addressed review comments
- Reordered items in tablets module a bit for consistency
- Custom implementation of PartialEq for TableTablets (because we want to compare Arc<Node> using Arc::ptr_eq)
- PartialEq / Eq impls in tablets module only for tests

Lorak-mmk commented:
I did try to perform a local test (usingcql-stress) to see tablet usage in monitoring, but didn't succeed.
I setup a 3-node scylla cluster:
```
ccm create tablets -i 127.0.1. -n 3 --scylla -v 'unstable/master:2024-05-01T18:26:10Z'
ccm updateconf 'experimental_features: [consistent-topology-changes, tablets]'
ccm start
```
and start monitoring:
```
./start-all.sh -v 6.0 -l
```
with `prometheus/scylla_servers.yml` looking like this:
```
- targets:
      - 127.0.1.1
      - 127.0.1.2
      - 127.0.1.3
  labels:
      cluster: cluster1
      dc: dc1
```

How it went:
- Sometimes when starting cql-stress I see a bump in non-token-aware queries in monitoring: 
![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/964beee7-fedc-4bd9-a2b4-aba4278bb20b)
The bump is very small: there are only a few tablets in the table, so they are being fetched in just a few queries (which is confirmed by driver tracing).
- I wanted to use tablet allocator shuffle API to force tablets to continously shuffle, but it doesn't seem to do anything. I call it like this:
`curl -X POST http://127.0.1.1:10000/v2/error_injection/injection/tablet_allocator_shuffle -d '{"one_shot": true}'` 
I get HTTP response 200, but driver doesn't receive any new tablet info (so presumably they are bot shuffled).

I don't think I'm going to see non-token-aware metric behave like I expected unless there are a lot of tablets, which is too problematic to do in a local test.
Still the behavior shown by driver logs confirms that tablet awareness works as expected, which is enough for me.

wprzytula commented on scylla/src/transport/locator/tablets.rs:
```diff
@@ -0,0 +1,1379 @@
+        self.tablet_list.insert(left_idx, tablet);
+    }
+
+    fn perform_maintanance(
+        &mut self,
+        removed_nodes: &HashSet<Uuid>,
+        all_current_nodes: &HashMap<Uuid, Arc<Node>>,
+        recreated_nodes: &HashMap<Uuid, Arc<Node>>,
+    ) {
```
maintenance

Lorak-mmk commented:
SCT run didn't pass, but failure is most likely unrelated to this PR (according to @fruch it's caused by https://github.com/scylladb/scylla-cluster-tests/issues/7320 ).

wprzytula reviewed (approved):
Impressive work! Well done!

fruch commented:
> SCT run didn't pass, but failure is most likely unrelated to this PR (according to @fruch it's caused by [scylladb/scylla-cluster-tests#7320](https://github.com/scylladb/scylla-cluster-tests/issues/7320) ).

just a general comment, I would recommend putting links to jenkins or Argus when you are referring specific SCT runs
it would help to go back to that results, or spin up the monitor to see specific things (i.e. to review)

Lorak-mmk commented:
> > SCT run didn't pass, but failure is most likely unrelated to this PR (according to @fruch it's caused by [scylladb/scylla-cluster-tests#7320](https://github.com/scylladb/scylla-cluster-tests/issues/7320) ).
> 
> just a general comment, I would recommend putting links to jenkins or Argus when you are referring specific SCT runs it would help to go back to that results, or spin up the monitor to see specific things (i.e. to review)


Sure, here's a run I'm talking about: https://jenkins.scylladb.com/job/scylla-staging/job/karol_baryla/job/longevity-100gb-4h-cql-stress-test-tablets/6/

Lorak-mmk commented:
One last small change: added a debug log in `TabletsInfo::add_tablet`, so that something about tablets is visible in `DEBUG` level, I think it will aid debugging in the future.

Lorak-mmk commented:
Fixed flaky integration test. Now the PR is finally ready to merge!

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,601 @@
+            // running with 3-node cluster and keyspace has replication factor 2,
+            // so even if each node has only one shard there must exist at least
+            // one non-replica node.
+            // Initially there are not tablets, so LBP will fallback to token ring.
```
there are no tablets

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,601 @@
+            // so even if each node has only one shard there must exist at least
+            // one non-replica node.
+            // Initially there are not tablets, so LBP will fallback to token ring.
+            // Tablet replicas and toklen ring replicas may overlap sometimes,
```
token ring

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,601 @@
+            // one non-replica node.
+            // Initially there are not tablets, so LBP will fallback to token ring.
+            // Tablet replicas and toklen ring replicas may overlap sometimes,
+            // so we can't guarantee a reply for each tablet, until we get any,
+            // so that LBP knows that table is a tablet table.
```
I don't get this. Until we get any what? Please rephrase it more clearly.

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,601 @@
+    rx: &mut mpsc::UnboundedReceiver<(ResponseFrame, Option<TargetShard>)>,
+) -> usize {
+    std::iter::from_fn(|| rx.try_recv().ok())
+        .map(|(frame, _shard)| frame_has_tablet_feedback(frame))
+        .filter(|b| *b)
```
nit: why not `filter_map` ?

Lorak-mmk replied:
`filter_map` expects `Option`, not `bool`, this would require code like `.filter_map(|(frame, _shard)| frame_has_tablet_feedback(frame).then_some(()))`. I really don't like this `.then_some(())` call on `bool`. Alternative is explicit match, but that's a lot of code.
map + filter still seems to me like the easiest option here.

wprzytula commented on scylla/tests/integration/tablets.rs:
```diff
@@ -0,0 +1,601 @@
+                try_join_all((0..100).map(|_| async { session.execute(&prepared, values).await }))
+                    .await
+                    .unwrap();
+                let feedbacks: usize = feedback_rxs.iter_mut().map(count_tablet_feedbacks).sum();
+                if feedbacks > 0 {
+                    total_tablets_with_feedback += 1;
+                }
```
When are we going to have `feedbacks > 1`? Are we sure all those feedbacks come from the same tablet?

Lorak-mmk commented:
Rewrote the whole comment mentioned in the review from scratch - so I marked the review comments regarding it as solved, as they are no longer relevant.
